zio-cache/shared/src/main/scala/zio/cache/Cache.scala

```diff
@@ -1,226 +1,236 @@
-                  .modify[IO[E, Value]] { state =>
-                    val entryStats = state.entryStats
-                    val map        = state.map
-
-                    map.get(key) match {
-                      case Some(MapValue.Pending(promise)) =>
-                        (trackHit(now, key) *> restore(promise.await), state)
-                      case Some(MapValue.Complete(exit)) =>
-                        (trackHit(now, key) *> IO.done(exit), state)
-                      case None =>
-                        val promise = Promise.unsafeMake[E, Value](fiberId)
-
-                        val lookupValue: UIO[Exit[E, Value]] =
-                          restore(lookup(key)).run
-                            .flatMap(exit => promise.done(exit).flatMap(_ => ZIO.succeedNow(exit)))
-                            .provide(env)
-
-                        val stats = entryStats.getOrElse(key, EntryStats.make(now))
-
-                        (
-                          trackMiss.flatMap(_ =>
-                            lookupValue
-                              .flatMap(exit => recordEntry(ref, now, key, stats, exit))
-                          ),
-                          state.copy(
-                            map = addAndPrune(state, key, promise),
-                            entryStats = entryStats + (key -> stats)
-                          )
-                        )
-                    }
+        new Cache[Key, Error, Value] {
+
+          def get(k: Key): IO[Error, Value] =
+            ZIO.effectSuspendTotal {
+              var key: MapKey[Key]               = null
+              var promise: Promise[Error, Value] = null
+              var value                          = map.get(k)
+              if (value eq null) {
+                promise = Promise.unsafeMake[Error, Value](fiberId)
+                key = new MapKey(k)
+                value = map.putIfAbsent(k, MapValue.Pending(key, promise))
+              }
+              if (value eq null) {
+                trackAccess(key)
+                trackMiss()
+                lookup(k)
+                  .provide(environment)
+                  .run
+                  .flatMap { exit =>
+                    val now        = Instant.now()
+                    val entryStats = EntryStats(now)
+
+                    map.put(k, MapValue.Complete(key, exit, entryStats))
+                    promise.done(exit) *> ZIO.done(exit)
                   }
-                  .flatMap(identityFn)
+                  .onInterrupt(promise.interrupt *> ZIO.succeed(map.remove(k)))
+              } else {
+                value match {
+                  case MapValue.Pending(key, promise) =>
+                    trackAccess(key)
+                    trackHit()
+                    promise.await
+                  case MapValue.Complete(key, exit, entryState) =>
+                    trackAccess(key)
+                    trackHit()
+                    val now  = Instant.now()
```

Comment:
This is fine but it almost makes me think we should have a `.unsafe` on core services that yields a package private Clock.Unsafe`, e.g.